File: common/platform-dotnet/SoundMetrics.Common/NetworkInterfaceInfo.fs
// Copyright 2019 Sound Metrics Corp. All Rights Reserved.

namespace SoundMetrics.Common

open System.Net
open System.Net.NetworkInformation
open System.Net.Sockets

[<AutoOpen>]
module private NetworkInterfaceInfoDetails =

    let NoAddress = lazy ( IPAddress 0L )
    let NoIndexFound = -1

    // Acknowledge that the IP properties may be null
    let toValueOption<'T> (t : 'T) : 'T voption =

        if isNull (box t) then
            ValueNone
        else
            ValueSome t

[<Struct>]
type NetworkInterfaceInfo = {
    /// The interface "index" as retrieved via
    /// ifc.GetIPProperties().GetIPv*Properties().Index
    Index:      int

    Name:       string
    Interface:  NetworkInterface
}
with
    static member NoIndex = NoIndexFound
    static member FromNetworkInterface(ifc : NetworkInterface) =

        let name = ifc.Name
        let props = ifc.GetIPProperties()

        let index =
            match props.GetIPv4Properties() |> toValueOption with
            | ValueSome v4 -> v4.Index
            | ValueNone ->
                match props.GetIPv6Properties() |> toValueOption with
                | ValueSome v6 -> v6.Index
                | ValueNone -> NetworkInterfaceInfo.NoIndex

        { Index = index; Name = name; Interface = ifc }

    static member internal Mask(a: byte array, mask: byte array) =

                if a.Length <> mask.Length then
                    invalidArg "c" "Mismatched arguments"

                let output = Array.zeroCreate<byte> a.Length

                for i = 0 to a.Length - 1 do
                    output.[i] <- byte (a.[i] &&& mask.[i])

                output

    static member internal Mask(a: IPAddress, mask: IPAddress) =

            if a.AddressFamily <> AddressFamily.InterNetwork then
                invalidArg "a" (sprintf "Only IPv4 is supported: %A" a)

            let xored = NetworkInterfaceInfo.Mask(a.GetAddressBytes(), mask.GetAddressBytes())
            IPAddress(xored)

    static member internal IsReachable(addr: IPAddress, subnetAddress: IPAddress, ipv4Mask: IPAddress) =
            let a = NetworkInterfaceInfo.Mask(addr, ipv4Mask)
            let b = NetworkInterfaceInfo.Mask(subnetAddress, ipv4Mask)
            a = b

    static member IsAddressReachable(addr: IPAddress,
                                     ifc: NetworkInterface) =

        if addr.AddressFamily = AddressFamily.InterNetwork then
            let props = ifc.GetIPProperties()

            let matchCount =
                props.UnicastAddresses
                |> Seq.filter (fun ua ->
                    ua.Address.AddressFamily = AddressFamily.InterNetwork
                        &&  NetworkInterfaceInfo.IsReachable(addr, ua.Address, ua.IPv4Mask))
                |> Seq.length
            matchCount > 0
        elif addr.AddressFamily = AddressFamily.InterNetworkV6 then
            true
        else
            false

File: common/platform-dotnet/SoundMetrics.Aris/Network/NetworkInterfaceInfo.cs
// Copyright 2019-2020 Sound Metrics Corp. All Rights Reserved.

using System;
using System.Linq;
using System.Net;
using System.Net.NetworkInformation;
using System.Net.Sockets;

namespace SoundMetrics.Aris.Network
{
    public struct NetworkInterfaceInfo
    {
        /// The interface "index" as retrieved via
        /// ifc.GetIPProperties().GetIPv*Properties().Index
        public int Index;

        public string Name;
        public NetworkInterface Interface;

        public static int NoIndex = -1;

        public static NetworkInterfaceInfo FromNetworkInterface(NetworkInterface ifc)
        {
            if (ifc is null)
            {
                throw new ArgumentNullException(nameof(ifc));
            }

            var name = ifc.Name;
            var props = ifc.GetIPProperties();

            int index = NoIndex;

            if (props.GetIPv4Properties() is IPv4InterfaceProperties v4)
            {
                index = v4.Index;
            }
            else if (props.GetIPv6Properties() is IPv6InterfaceProperties v6)
            {
                index = v6.Index;
            }

            return new NetworkInterfaceInfo
            {
                Index = index,
                Name = name,
                Interface = ifc,

            };
        }

        internal static byte[] Mask(byte[] a, byte[] mask)
        {
            if (a.Length != mask.Length)
            {
                throw new ArgumentException("Mismatched argument lengths", nameof(a));
            }

            var output = new byte[a.Length];

            for (int i = 0; i < a.Length - 1; ++i)
            {
                output[i] = (byte)(a[i] & mask[i]);
            }

            return output;
        }

        internal static IPAddress Mask(IPAddress a, IPAddress mask)
        {
            if (a.AddressFamily != AddressFamily.InterNetwork)
            {
                throw new ArgumentException(
                    $"Only IPv4 is supported: {a}");
            }

            var xored = Mask(a.GetAddressBytes(), mask.GetAddressBytes());
            var masked = new IPAddress(xored);
            return masked;
        }

        internal static bool IsReachable(
            IPAddress addr,
            IPAddress subnetAddress,
            IPAddress ipv4Mask)
        {
            var maskedAddr = Mask(addr, ipv4Mask);
            var maskedSubnet = Mask(subnetAddress, ipv4Mask);
            var areEqual = maskedAddr.Equals(maskedSubnet);
            return areEqual;
        }

        internal static bool IsAddressReachable(
            IPAddress addr,
            NetworkInterface ifc)
        {
            if (addr.AddressFamily == AddressFamily.InterNetwork)
            {
                var props = ifc.GetIPProperties();
                var matchCount =
                    props.UnicastAddresses
                        .Count(ua =>
                            ua.Address.AddressFamily == AddressFamily.InterNetwork
                            && IsReachable(addr, ua.Address, ua.IPv4Mask)
                        );

                return matchCount > 0;
            }
            else if (addr.AddressFamily == AddressFamily.InterNetworkV6)
            {
                return true;
            }
            else
            {
                return false;
            }
        }
    }
}
